### PR TITLE
CFE-3164: Added MAC address to sys.hardware_(addresses,mac)

### PR DESCRIPTION
**Pull request by oleorhagen**
Previously MAC addresses would not be added to the sys.hardware_addresses
and the sys.hardware_mac variables if the interface contained only IPv6
addresses.

Changelog: Commit

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a pull request! Maybe @vpodzime can review this?

File: libenv/unix_iface.c [hunk 1]
```diff
@@ -677,6 +675,45 @@ static void FindV6InterfacesInfo(EvalContext *ctx, Rlist **interfaces, Rlist **h
             }
         }
 
+        if (strcasestr(ifconfig_line, "ether"))
```
Comment (olehermanse) [inline libenv/unix_iface.c]: **Better:**

```
if (strcasestr(ifconfig_line, "ether") != NULL)
```

**Best:**

```
if (StringStartsWith(TrimWhitespace(ifconfig_line), "ether"))
```

File: libenv/unix_iface.c [hunk 2]
```diff
@@ -677,6 +675,45 @@ static void FindV6InterfacesInfo(EvalContext *ctx, Rlist **interfaces, Rlist **h
...
+        {
+            char *hw_mac = strtok(ifconfig_line, " ");
+            if (hw_mac == NULL) {
+              Log(LOG_LEVEL_ERR, "Failed to parse hw_mac address");
+            }
+            hw_mac = strtok(NULL, " ");
+            if (hw_mac == NULL) {
+              Log(LOG_LEVEL_ERR, "Failed to parse hw_mac address");
+            }
```
Comment (olehermanse) [inline libenv/unix_iface.c]: [`strtok` is not particularly readable or safe](http://man7.org/linux/man-pages/man3/strtok.3.html#BUGS). I'd prefer something like `SeqStringFromString()`:

```C
Seq *const ether_line = SeqStringFromString(ifconfig_line, " ");
const size_t length = SeqLength(ether_line);

if (length < 2 || !StringSafeEqualIgnoreCase(SeqAt(ether_line, 0), "ether"))
{
    Log(LOG_LEVEL_ERR, "Failed to parse hw_mac address");
}

const char *const hw_mac = SeqAt(ether_line, 1);

if (!RlistContainsString(*hardware, hw_mac))
{
    [...]
}
SeqDestroy(ether_line);
```

File: libenv/unix_iface.c [hunk 3]
```diff
@@ -677,6 +675,45 @@ static void FindV6InterfacesInfo(EvalContext *ctx, Rlist **interfaces, Rlist **h
...
+                      "hardware_mac[%s]",
+                      CanonifyName(current_interface));
+
+                  EvalContextVariablePutSpecial(
+                      ctx,
+                      SPECIAL_SCOPE_SYS,
+                      name,
+                      hw_mac,
+                      CF_DATA_TYPE_STRING,
+                      "source=agent");
```
Comment (olehermanse) [inline libenv/unix_iface.c]: I would add another tag:

```
"source=agent,derived-from=ifconfig"
```

File: libenv/unix_iface.c (hunk 1 above)
Comment (olehermanse) [reply libenv/unix_iface.c]: We could add `StringStartsWithIgnoreCase()`. Have you seen `ether` uppercased somewhere?

Comment (oleorhagen) [reply libenv/unix_iface.c]: Nope, but why not :+1:

File: libenv/unix_iface.c (hunk 2 above)
Comment (oleorhagen) [reply libenv/unix_iface.c]: I knew there was a comment inbound for this one :p

**Comment by oleorhagen** [comment]
Without changes:

```
root@dev core $ cf-agent --verbose -B 192.168.100.10 > t.out
root@dev core $ cf-promises --show-vars | grep sys.hardware
default:host_info_report_inventory.inventory_vars {"default:cfe_autorun_inventory_dmidecode.total_physical_memory_MB":"","default:cfe_autorun_inventory_ipv4_addresses.ipv4[10.0.2.15]":"10.0.2.15","default:cfe_autorun_inventory_ipv4_addresses.ipv4[192.168.100.10]":"192.168.100.10","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fde4:8dba:82e1::c4]":"fde4:8dba:82e1::c4","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fe80::a00:27ff:fe1a:fd8b]":"fe80::a00:27ff:fe1a:fd8b","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fe80::a00:27ff:fed2:5fbf]":"fe80::a00:27ff:fed2:5fbf","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fe80::b7:5cff:fe02:92a2]":"fe80::b7:5cff:fe02:92a2","default:cfe_autorun_inventory_listening_ports.ports":["22","53","5308"],"default:inventory_any.id":"b15bbb63bd4f95f15e528e11d7a4c10d3d0b6b7e","default:inventory_os.description":"ubuntu_18 (LSB missing)","default:sys.arch":"x86_64","default:sys.cf_version":"3.15.0a.ce80dd03d","default:sys.class":"linux","default:sys.cpus":"4","default:sys.flavor":"ubuntu_18","default:sys.fqhost":"dev","default:sys.hardware_addresses":["02:b7:5c:02:92:a2","08:00:27:1a:fd:8b"],"default:sys.host":"dev","default:sys.interfaces":["enp0s3","enp0s8","enp0s9"],"default:sys.ipv4":"10.0.2.15","default:sys.key_digest":"MD5=4a086ac40098e2f6f169f459b66cce81","default:sys.os_release":{"BUG_REPORT_URL":"https://bugs.launchpad.net/ubuntu/","HOME_URL":"https://www.ubuntu.com/","ID":"ubuntu","ID_LIKE":"debian","NAME":"Ubuntu","PRETTY_NAME":"Ubuntu 18.04.3 LTS","PRIVACY_POLICY_URL":"https://www.ubuntu.com/legal/terms-and-policies/privacy-policy","SUPPORT_URL":"https://help.ubuntu.com/","UBUNTU_CODENAME":"bionic","VERSION":"18.04.3 LTS (Bionic Beaver)","VERSION_CODENAME":"bionic","VERSION_ID":"18.04"},"default:sys.release":"4.15.0-65-generic","default:sys.uptime":"339","default:sys.uqhost":"dev"} source=promise                          
default:sys.hardware_addresses            {"02:b7:5c:02:92:a2","08:00:27:1a:fd:8b"}                   inventory,source=agent,attribute_name=MAC addresses
default:sys.hardware_flags                {"up loopback running","up broadcast running multicast","up broadcast running multicast"} source=agent                            
default:sys.hardware_mac[enp0s3]         02:b7:5c:02:92:a2                                            source=agent                            
default:sys.hardware_mac[enp0s8]         08:00:27:1a:fd:8b                                            source=agent 
```


With changes:

```  
root@dev core $ cf-promises --show-vars | grep sys.hardware
default:host_info_report_inventory.inventory_vars {"default:cfe_autorun_inventory_dmidecode.total_physical_memory_MB":"","default:cfe_autorun_inventory_ipv4_addresses.ipv4[10.0.2.15]":"10.0.2.15","default:cfe_autorun_inventory_ipv4_addresses.ipv4[192.168.100.10]":"192.168.100.10","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fde4:8dba:82e1::c4]":"fde4:8dba:82e1::c4","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fe80::a00:27ff:fe1a:fd8b]":"fe80::a00:27ff:fe1a:fd8b","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fe80::a00:27ff:fed2:5fbf]":"fe80::a00:27ff:fed2:5fbf","default:cfe_autorun_inventory_ipv4_addresses.ipv4[fe80::b7:5cff:fe02:92a2]":"fe80::b7:5cff:fe02:92a2","default:cfe_autorun_inventory_listening_ports.ports":["22","53","5308"],"default:inventory_any.id":"b15bbb63bd4f95f15e528e11d7a4c10d3d0b6b7e","default:inventory_os.description":"ubuntu_18 (LSB missing)","default:sys.arch":"x86_64","default:sys.cf_version":"3.15.0a.ce80dd03d","default:sys.class":"linux","default:sys.cpus":"4","default:sys.flavor":"ubuntu_18","default:sys.fqhost":"dev","default:sys.hardware_addresses":["02:b7:5c:02:92:a2","08:00:27:1a:fd:8b","08:00:27:d2:5f:bf"],"default:sys.host":"dev","default:sys.interfaces":["enp0s3","enp0s8","enp0s9"],"default:sys.ipv4":"10.0.2.15","default:sys.key_digest":"MD5=4a086ac40098e2f6f169f459b66cce81","default:sys.os_release":{"BUG_REPORT_URL":"https://bugs.launchpad.net/ubuntu/","HOME_URL":"https://www.ubuntu.com/","ID":"ubuntu","ID_LIKE":"debian","NAME":"Ubuntu","PRETTY_NAME":"Ubuntu 18.04.3 LTS","PRIVACY_POLICY_URL":"https://www.ubuntu.com/legal/terms-and-policies/privacy-policy","SUPPORT_URL":"https://help.ubuntu.com/","UBUNTU_CODENAME":"bionic","VERSION":"18.04.3 LTS (Bionic Beaver)","VERSION_CODENAME":"bionic","VERSION_ID":"18.04"},"default:sys.release":"4.15.0-65-generic","default:sys.uptime":"337","default:sys.uqhost":"dev"} source=promise                          
default:sys.hardware_addresses            {"02:b7:5c:02:92:a2","08:00:27:1a:fd:8b","08:00:27:d2:5f:bf"} inventory,source=agent,attribute_name=MAC addresses
default:sys.hardware_flags                {"up loopback running","up broadcast running multicast","up broadcast running multicast"} source=agent                            
default:sys.hardware_mac[enp0s3]         02:b7:5c:02:92:a2                                            source=agent                            
default:sys.hardware_mac[enp0s8]         08:00:27:1a:fd:8b                                            source=agent                            
default:sys.hardware_mac[enp0s9]         08:00:27:d2:5f:bf                                            source=agent,derived-from=ifconfig      
```

**Comment by oleorhagen** [comment]
I don't think the jenkins errors are my fault

**Comment by oleorhagen** [comment]
```
root@dev core $ ip a s
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 02:b7:5c:02:92:a2 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s3
       valid_lft 64246sec preferred_lft 64246sec
    inet6 fe80::b7:5cff:fe02:92a2/64 scope link 
       valid_lft forever preferred_lft forever
3: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 08:00:27:1a:fd:8b brd ff:ff:ff:ff:ff:ff
    inet 192.168.100.10/24 brd 192.168.100.255 scope global enp0s8
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe1a:fd8b/64 scope link 
       valid_lft forever preferred_lft forever
4: enp0s9: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 08:00:27:d2:5f:bf brd ff:ff:ff:ff:ff:ff
    inet6 fde4:8dba:82e1::c4/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fed2:5fbf/64 scope link 
       valid_lft forever preferred_lft forever
```

File: libenv/unix_iface.c [hunk 4]
```diff
@@ -677,6 +676,57 @@ static void FindV6InterfacesInfo(EvalContext *ctx, Rlist **interfaces, Rlist **h
...
+        const char *const stripped_ifconfig_line =
+            TrimWhitespace(ifconfig_line);
+        if (StringStartsWith(stripped_ifconfig_line, "ether"))
+        {
+            Seq *const ether_line =
+                SeqStringFromString(stripped_ifconfig_line, ' ');
+            const size_t length = SeqLength(ether_line);
+
+            if (length < 2
+                || !StringSafeEqual_IgnoreCase(SeqAt(ether_line, 0), "ether"))
```
Comment (vpodzime) [inline libenv/unix_iface.c]: If you changed the above `StringStartsWith()` to check for `ether ` (including the space), you'd be able to drop this second part of the condition here.

File: libenv/unix_iface.c [hunk 5]
```diff
@@ -677,6 +676,57 @@ static void FindV6InterfacesInfo(EvalContext *ctx, Rlist **interfaces, Rlist **h
...
+
+                if (!RlistContainsString(*hardware, hw_mac))
+                {
+                    Log(LOG_LEVEL_VERBOSE,
+                        "Adding MAC address: %s for %s",
+                        hw_mac,
+                        current_interface);
+
+                    RlistAppend(hardware, hw_mac, RVAL_TYPE_SCALAR);
+                    assert(RlistContainsString(*hardware, hw_mac));
```
Comment (vpodzime) [inline libenv/unix_iface.c]: Please move this assert into the `RlistAppend()` function.

Comment (oleorhagen) [reply libenv/unix_iface.c]: If so, then should there be a `RlistAppendString` function ?

File: libenv/unix_iface.c [hunk 6]
```diff
@@ -677,6 +676,57 @@ static void FindV6InterfacesInfo(EvalContext *ctx, Rlist **interfaces, Rlist **h
...
+
+                    char name[CF_MAXVARSIZE];
```
Comment (olehermanse) [inline libenv/unix_iface.c]: `variable_name` is a more clear name, imo.

**Comment by olehermanse** [comment]
@cf-bottom, build this in jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2928/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2928/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2928/

File: libenv/unix_iface.c (hunk 5 above)
Comment (vpodzime) [reply libenv/unix_iface.c]: Sounds like a plan to me!

Comment (vpodzime) [reply libenv/unix_iface.c]: But that sounds like something that's a bit out of scope of this PR>

**Comment by olehermanse** [comment]
@cf-bottom jenkins again, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2980/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2980/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2980/

**Pull request merged**